### PR TITLE
OSOE-47: Add VerifyElementTexts extension, interface for PageNavigationState, EnableApplicationInsightsOfflineOperation() extension method

### PR DESCRIPTION
**Pull request by sarahelsaig**
OSOE-47

### CODE REVIEW

File: Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj [hunk 1]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <IsPackable>false</IsPackable>
+    <ErrorOnDuplicatePublishOutputFiles>false</ErrorOnDuplicatePublishOutputFiles>
```
Comment (Piedone) [inline Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: Why is this needed?

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs [hunk 2]
```diff
@@ -101,6 +104,38 @@ public static void SucccessMessageExists(this UITestContext context, string matc
                 .Text
                 .ShouldBe(errorMessage);
 
+        /// <summary>
+        /// Retrieves the elements according to <paramref name="by"/> and matches their text content against <paramref
+        /// name="toMatch"/>. The strings are trimmed in both collections. If an item in <paramref name="toMatch"/> is
```
Comment (Piedone) [inline Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs]: *both* collections?

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs [hunk 3]
```diff
@@ -101,6 +104,38 @@ public static void SucccessMessageExists(this UITestContext context, string matc
...
+                .ToArray();
+
+            context
+                .GetAll(by)
+                .Select((element, index) => dontCare.Contains(index) ? null : element.Text.Trim())
+                .ToArray()
+                .ShouldBe(target);
+        }
+
+        /// <inheritdoc cref="VerifyElementTexts(Lombiq.Tests.UI.Services.UITestContext,OpenQA.Selenium.By,object[])"/>
```
Comment (Piedone) [inline Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs]: image

File: Lombiq.Tests.UI/Models/IWebContentState.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /// <summary>
+    /// Represents a state of some web content that may change, for example via page navigation or dynamic content after
+    /// user interaction.
+    /// </summary>
+    public interface IWebContentState
+    {
+        /// <summary>
+        /// Returns <see langword="true"/> if navigation has occurred or the content has changed based on some
+        /// previously provided content..
```
Comment (Piedone) [inline Lombiq.Tests.UI/Models/IWebContentState.cs]:
```suggestion
        /// previously provided content.
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj (hunk 1 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: ["Publish error: Found multiple publish output files with the same relative path"](https://github.com/Lombiq/UI-Testing-Toolbox/pull/132/commits/bbb18b4a294b6dd02fd11ac5f75397916d949f64)

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs (hunk 2 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs]: The results and toMatch are both trimmed. This is because if you accidentally left a space in the toMatch it would be hell to debug.

File: Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj (hunk 1 above)
Comment (Piedone) [reply Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: Where is that happening though? In CI (because I can't see it)?

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs (hunk 2 above)
Comment (Piedone) [reply Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs]: I see. It was unclear to me what the other collection is. So maybe:

```c#
        /// Retrieves the elements according to <paramref name="by"/> and matches their text content against <paramref
        /// name="toMatch"/>. Both the text contents and <paramref name="toMatch"/> strings are trimmed.. If an item in
        /// <paramref name="toMatch"/> is <see langword="null" /> it's ignored among the result elements too. Every
        /// other item is converted to string, using invariant culture where possible.
```

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs [hunk 5]
```diff
@@ -101,6 +104,38 @@ public static void SucccessMessageExists(this UITestContext context, string matc
                 .Text
                 .ShouldBe(errorMessage);
 
+        /// <summary>
+        /// Retrieves the elements according to <paramref name="by"/> and matches their text content against <paramref
+        /// name="toMatch"/>. The strings are trimmed in both collections. If an item in <paramref name="toMatch"/> is
+        /// <see langword="null" /> it's ignored among the result elements too. Every other item is converted to string,
+        /// using invariant culture where possible.
+        /// </summary>
+        public static void VerifyElementTexts(this UITestContext context, By by, params object[] toMatch)
```
Comment (Piedone) [inline Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs:112]: Why is `toMatch` not `string[]`?

File: Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj (hunk 1 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: While trying to fix [this](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_PullAndBuild/950044?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true). I reproduced it locally, fixed it and then I got the above error when rerunning the same command.

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs (hunk 5 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs:112]: In case you want to compare numbers too.

File: Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj (hunk 1 above)
Comment (Piedone) [reply Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: Didn't you fix this centrally, @BenedekFarkas?

File: Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs (hunk 5 above)
Comment (Piedone) [reply Lombiq.Tests.UI/Extensions/ElementRetrievalUITestContextExtensions.cs:112]: I see, OK then.

File: Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj (hunk 1 above)
Comment (BenedekFarkas) [reply Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: Locally you need it of course, but in TeamCity, this parameter is passed into every `dotnet publish` command we run (regardless of calling from PS or using the .NET Runner).
Btw https://docs.microsoft.com/en-us/dotnet/core/compatibility/sdk/6.0/duplicate-files-in-output

Comment (Piedone) [reply Lombiq.Tests.UI.Samples/Lombiq.Tests.UI.Samples.csproj:8]: OK, so this is not something that would cause any issues in TC. It won't hurt locally though but Dávid please document why this is necessary with this page linked.

File: Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs [hunk 6]
```diff
@@ -161,6 +161,7 @@ public void AssertBrowserLogMaybe(IList<BrowserLogMessage> browserLogs, Action<s
                 !message.Message.ContainsOrdinalIgnoreCase("HTML Imports is deprecated") &&
                 // The 404 is because of how browsers automatically request /favicon.ico even if a favicon is declared
                 // to be under a different URL.
-                !message.IsNotFoundMessage("/favicon.ico");
+                !message.IsNotFoundMessage("/favicon.ico") &&
+                !message.Message.ContainsOrdinalIgnoreCase("https://dc.services.visualstudio.com/v2/track");
```
Comment (Piedone) [inline Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: Why is this necessary?

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs (hunk 6 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: I'm not sure. It might be just becuase I'm not using VS. Let me remove it and see if it pops up in CI.

Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: Yup, tests passed without it, so I guess it can be safely ignored.

Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: [Ohno, it happened again](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_RunUiTests/954720?showRootCauses=false&expandBuildChangesSection=true&expandBuildTestsSection=true) in OSOE-64. This time in CI. [It seems to related to Application Insights](https://stackoverflow.com/questions/49148355/what-is-the-use-of-https-dc-services-visualstudio-com-v2-track), is that set up correctly?

Comment (Piedone) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: The AI module should be configured with offline mode, see `UITestBase` in the Samples.

**Review by Piedone — APPROVED** [review]

File: Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs [hunk 7]
```diff
@@ -161,6 +161,9 @@ public void AssertBrowserLogMaybe(IList<BrowserLogMessage> browserLogs, Action<s
                 !message.Message.ContainsOrdinalIgnoreCase("HTML Imports is deprecated") &&
                 // The 404 is because of how browsers automatically request /favicon.ico even if a favicon is declared
                 // to be under a different URL.
-                !message.IsNotFoundMessage("/favicon.ico");
+                !message.IsNotFoundMessage("/favicon.ico") &&
+                // Ignore if browser telemetry is not available.
+                !message.Message.ContainsOrdinalIgnoreCase(
+                    "https://dc.services.visualstudio.com/v2/track - Failed to load resource: the server responded with a status of 400");
```
Comment (Piedone) [inline Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: Why do we have this back? As I mentioned before, the AI module should be configured with offline mode, see `UITestBase` in the Samples. Such requests should never happen during UI testing.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs (hunk 7 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: Ah, I misunderstood. I thought you meant that it's already configured like that so it can't happen.

Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: By the way is there any reason AI isn't offline by default in the UI tests, or there isn't at least an ~extension~ method for turning it off? 
(EDIT: regular method because "Either move this extension to another namespace or move the method inside the class itself.")

Comment (Piedone) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: For both we need something related to UI testing in the AI module that we don't have yet but I added it here: https://github.com/Lombiq/Orchard-Azure-Application-Insights/issues/25

Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: [Added the method!](https://github.com/Lombiq/UI-Testing-Toolbox/pull/132/commits/e81c679df9eaa1554c50a2e3b743a34c58ab2014#diff-b1534dd6c730fb4e0c76a1498e1f41e0af19b8f6e79d4c9dea5d1906257de8c0R33)

File: Lombiq.Tests.UI/Services/OrchardCoreInstance.cs [hunk 8]
```diff
@@ -24,6 +24,27 @@ public class OrchardCoreConfiguration
...
+        /// </summary>
+        public void EnableApplicationInsightsOfflineOperation() =>
+            BeforeAppStart +=
+                (_, argumentsBuilder) =>
+                {
+                    // This is quite handy! We're adding a configuration parameter when launching the app. This
+                    // can be used to set configuration for configuration providers, see the docs:
+                    // https://docs.orchardcore.net/en/latest/docs/reference/core/Configuration/.
+                    // What's happening here is that we set the Lombiq Application Insights module's parameter
+                    // to allow us to test it. We'll get back to this later when writing the actual test.
```
Comment (Piedone) [inline Lombiq.Tests.UI/Services/OrchardCoreInstance.cs]: Keep this in the sample, please.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs (hunk 7 above)
Comment (Piedone) [reply Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: Nice!

File: Lombiq.Tests.UI/Services/OrchardCoreInstance.cs (hunk 8 above)
Comment (sarahelsaig) [reply Lombiq.Tests.UI/Services/OrchardCoreInstance.cs]: The whole comment?

Comment (Piedone) [reply Lombiq.Tests.UI/Services/OrchardCoreInstance.cs]: Yes.

**Review by Piedone — APPROVED** [review]
